### PR TITLE
Fixed #30841 -- Deprecated using non-boolean values for isnull lookup.

### PR DESCRIPTION
**Pull request by aericson**
Add deprecation warning to, later on, add a check on version 4.x

### CODE REVIEW

File: django/db/models/lookups.py [hunk 1]
```diff
@@ -463,6 +465,12 @@ class IsNull(BuiltinLookup):
     prepare_rhs = False
 
     def as_sql(self, compiler, connection):
+        if not isinstance(self.rhs, bool):
+            warnings.warn(
+                'Providing a non-bool value with a isnull lookup is untested and '
+                'undocumented leading to inconsistent behaviour.',
```
Comment (felixxm) [inline django/db/models/lookups.py]: Maybe:
```
'Using a non-boolean value with an isnull lookup is deprecated, '
'use True or False.'
```
I don't think that we need to modify `stacklevel`. We should also add a comment that will make easier to remove this in Django 4.0, e.g.
```python
if not isinstance(self.rhs, bool):
    # When the deprecation ends, replace with:
    # raise ValueError(
    #     'The QuerySet value for an isnull lookup must be True or '
    #     'False.'
    # )
    warnings.warn(
        'Using a non-boolean value with an isnull lookup is '
        'deprecated, use True or False instead.',
        RemovedInDjango40Warning,
    )
```

File: docs/releases/3.1.txt [hunk 2]
```diff
@@ -258,6 +258,9 @@ Miscellaneous
 * ``PASSWORD_RESET_TIMEOUT_DAYS`` setting is deprecated in favor of
   :setting:`PASSWORD_RESET_TIMEOUT`.
 
+* Using ``isnull`` lookup with a non-boolean value is deprecated since it is
+  undocumented and untested and could lead to inconsistent behavior.
```
Comment (felixxm) [inline docs/releases/3.1.txt]:
```
* The undocumented usage of the ``isnull`` lookup with non-boolean values as
  the right-hand side is deprecated, use ``True`` or ``False`` instead.

File: tests/queries/tests.py [hunk 3]
```diff
@@ -3259,6 +3260,19 @@ def test_ticket_21748_complex_filter(self):
...
+            warnings.simplefilter('always')
+            ModelA.objects.filter(Q(b__name__isnull=1)).exists()
+
+        self.assertEqual(len(warns), 1)
+        msg = str(warns[0].message)
+        self.assertEqual(
+            msg,
+            'Providing a non-bool value with a isnull lookup is '
+            'untested and undocumented leading to inconsistent behaviour.'
+        )
```
Comment (felixxm) [inline tests/queries/tests.py]: Please move tests to the `lookup/tests.py`. Testing a deprecation warning can be simplified, e.g.
```python
    # These test will catch ValueError in Django 4.0 when using non-boolean
    # values with an isnull lookup becomes forbidden.

    def test_isnull_non_boolean_value(self):
        msg = (
            'Using a non-boolean value with an isnull lookup is deprecated, '
            'use True or False instead.'
        )
        with self.assertWarnsMessage(RemovedInDjango40Warning, msg):
            ModelA.objects.filter(b__name__isnull=1).exists()
```
We should add tests for all cases i.e. a normal field, related field, and related field with a multi-column relation (`ForeignObject`).

File: docs/internals/deprecation.txt [hunk 4]
```diff
@@ -36,6 +36,9 @@ details on these changes.
 
 * The ``PASSWORD_RESET_TIMEOUT_DAYS`` setting will be removed.
 
+* Passing a non-boolean value to the ``isnull`` lookup is no longer going
+  to be allowed.
```
Comment (felixxm) [inline docs/internals/deprecation.txt]:
```
* Usage of the ``isnull`` lookup with non-boolean values as the right-hand
  side will not longer be allowed.
```

**Comment by aericson** [comment]
> Should we also modify RelatedIsNull?

I've added tests for this but seems like [django/db/models/fields/related_lookups.py#L129](https://github.com/aericson/django/blob/58c7979685af2e846178f4d2f35b3eb530e44db2/django/db/models/fields/related_lookups.py#L129) will already trigger the warning.


I'm not 100% sure about the code on `RelatedLookupMixin.as_sql`. Should we also add the check there?

Thank you both for the review 💯

**Pull request merged**